### PR TITLE
Build and install refman with Dune.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure.

FTR `make -f Makefile.dune voboot && dune build coq-doc.install` works but `dune install coq-doc` gives an error:

```
Installing /tmp/lib/coq-doc/META
Installing /tmp/lib/coq-doc/dune-package
Installing /tmp/lib/coq-doc/opam
Installing /tmp/doc/coq-doc/LICENSE
Installing /tmp/doc/coq-doc/README.md
Installing /tmp/doc/coq-doc/html/refman
Error: exception Sys_error("Is a directory")
Backtrace:
Raised by primitive operation at file "stdlib.ml", line 408, characters 7-32
Called from file "src/dune/artifact_substitution.ml", line 378, characters 8-27
Called from file "src/fiber/fiber.ml", line 114, characters 10-15
Re-raised at file "src/stdune/exn.ml" (inlined), line 37, characters 27-56
Called from file "src/stdune/exn_with_backtrace.ml", line 13, characters 33-71
Called from file "src/fiber/fiber.ml", line 102, characters 8-15

I must not segfault.  Uncertainty is the mind-killer.  Exceptions are
the little-death that brings total obliteration.  I will fully express
my cases.  Execution will pass over me and through me.  And when it
has gone past, I will unwind the stack along its path.  Where the
cases are handled there will be nothing.  Only I will remain.
```

Probably a bug to report upstream. @ejgallego WDYT?

In any case, I'd like this PR to be merged without waiting too long. If you want, I can comment out the `install` stanza.

### CODE REVIEW

File: Makefile.dune [hunk 1]
```diff
@@ -32,6 +32,7 @@ help:
 	@echo ""
 	@echo "  - test-suite:  run Coq's test suite"
 	@echo "  - refman-html: build Coq's reference manual [HTML version]"
+	@echo "  - refman-html: build Coq's reference manual [PDF version]"
```
Comment (SkySkimmer) [inline Makefile.dune]:
```suggestion
	@echo "  - refman-pdf: build Coq's reference manual [PDF version]"
```

Comment (Zimmi48) [reply Makefile.dune]: Thanks!

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -418,14 +418,23 @@ doc:refman:
...
   extends: .dune-ci-template
   variables:
     DUNE_TARGET: refman-html
   artifacts:
     paths:
       - _build/log
-      - _build/default/doc/sphinx_build/html
+      - _build/default/doc/refman-html
+
+doc:refman-pdf:dune:
```
Comment (ejgallego) [inline .gitlab-ci.yml]: Do think it is worth the overhead to split the refman generation in two?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -446,11 +455,11 @@ doc:refman:deploy:
       - $DOCUMENTATION_DEPLOY_KEY
   dependencies:
     - doc:ml-api:odoc
-    - doc:refman:dune
+    - doc:refman-html:dune
```
Comment (ejgallego) [inline .gitlab-ci.yml]: IMHO not worth doing this renaming.

File: Makefile.dune [hunk 4]
```diff
@@ -32,6 +32,7 @@ help:
 	@echo ""
 	@echo "  - test-suite:  run Coq's test suite"
 	@echo "  - refman-html: build Coq's reference manual [HTML version]"
+	@echo "  - refman-pdf: build Coq's reference manual [PDF version]"
```
Comment (ejgallego) [inline Makefile.dune]: Nit on vertical alignment.

File: dev/dune [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 (rule
   (targets dune-dbg)
+  (package coq)
```
Comment (ejgallego) [inline dev/dune]: This seems unrelated, why do you want to install these objects ?

File: doc/dune [hunk 6]
```diff
@@ -10,23 +12,33 @@
...
  (targets unreleased.rst)
  (deps (source_tree changelog))
  (action (with-stdout-to %{targets} (bash "cat changelog/00-title.rst changelog/*/*.rst"))))
 
-; The install target still needs more work.
-; (install
-;  (section doc)
-;  (package coq-refman)
-;  (files sphinx_build))
+(install
```
Comment (ejgallego) [inline doc/dune]: Unfortunately this is broken [as you have witnessed] and still not fixed. The corresponding upstream issue is https://github.com/ocaml/dune/issues/1377 , see also https://github.com/coq/coq/pull/9636 and https://github.com/ocaml/dune/issues/1868

File: .gitlab-ci.yml (hunk 2 above)
Comment (Zimmi48) [reply .gitlab-ci.yml]: Basically, nothing is shared between the two targets, and they take quite long. I wouldn't want to delay the generation of the HTML artifact just to ensure that the PDF version builds. That's why I split this in two.

File: .gitlab-ci.yml (hunk 3 above)
Comment (Zimmi48) [reply .gitlab-ci.yml]: This is more consistent with the name of the targets though.

File: dev/dune (hunk 5 above)
Comment (Zimmi48) [reply dev/dune]: I don't want to install them. But without these additional `(package coq)`, I got errors when running `dune build -p coq-doc` like:

```
File "dev/shim/dune", line 28, characters 4-20:
28 |   %{bin:coqtop.byte}
         ^^^^^^^^^^^^^^^^
Error: Program coqtop.byte not found in the tree or in PATH
 (context: default)
```

File: doc/dune (hunk 6 above)
Comment (Zimmi48) [reply doc/dune]: I see, thanks for the links! But it seems that it is not just an issue of "installing directory is not supported" but also that Dune doesn't realize that this is unsupported and fails pretty badly when trying anyway.

File: dev/dune (hunk 5 above)
Comment (Zimmi48) [reply dev/dune]: Still, I reverted these changes.

**Comment by Zimmi48** [comment]
I kept the `install` rules because they are useful to make `coq-doc.install` an interesting target. Furthermore, the support is already partial with `_build/install/default/doc/coq-doc` containing exactly what is asked. See the updated comment.

File: .gitlab-ci.yml (hunk 2 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: I thought they would share some Sphinx databases, etc... The build didn't take long in my machine but it was using -j 4 tho.

File: .gitlab-ci.yml (hunk 3 above)
Comment (ejgallego) [reply .gitlab-ci.yml]: Not worth the conflict with future PRs IMO, the `dune` naming is gone after #8729 for example.

File: dev/dune (hunk 5 above)
Comment (ejgallego) [reply dev/dune]: Oh, this is due to `-p` assuming the dependencies to be already built; even so that error seems strange, let me see.

File: doc/dune (hunk 6 above)
Comment (ejgallego) [reply doc/dune]: Yes I agree; the solution to this is a bit tricky unfortunately. Some ideas are:
- generate the install file using a small program [but that's tricky to maintain]
- use a small makefile to install the docs [that's what dune does, well, in this case to deploy the sphinx manual as it is not distributed in the opam package I think]

File: coq-doc.opam [hunk 7]
```diff
@@ -27,7 +27,7 @@ build-env: [
 ]
 
 build: [
-  [ "dune" "build" "@refman" "-j" jobs ]
+  [ "dune" "build" "coq-doc.install" "-j" jobs ]
```
Comment (ejgallego) [inline coq-doc.opam]: This should be `"-p" "coq-doc"` , see upstream documentation https://dune.readthedocs.io/en/stable/opam.html

File: doc/dune [hunk 8]
```diff
@@ -10,23 +12,38 @@
...
+(rule
+ (targets refman-pdf)
+ (alias refman-pdf)
+ (package coq-doc)
+ (deps
+  (package coq)
+  (source_tree sphinx)
+  (source_tree tools/coqrst)
+  unreleased.rst
+  (env_var SPHINXWARNOPT))
```
Comment (ejgallego) [inline doc/dune]: Could we capture the common deps as an alias?

**Review by ejgallego — APPROVED** [review]
Looks mostly good, a few comments:
- we may want to avoid the renaming of the CI target as it should be undone soon in #8729 
- opam file should use the right dune invocation
- I'm unsure about uncommenting the install entries, they produce bogus install files so IMHO it is better to keep them disabled.

File: .gitlab-ci.yml (hunk 3 above)
Comment (Zimmi48) [reply .gitlab-ci.yml]: I see, thanks.

File: .gitlab-ci.yml (hunk 2 above)
Comment (Zimmi48) [reply .gitlab-ci.yml]: I tried to write the rules differently so that the cache would be shared but it didn't seem to work. Maybe there is something that I don't understand here though. I wonder if @cpitclaudel would have some insight for this.

**Comment by Zimmi48** [comment]
Updated to take your comments into account. I did not comment out the install rules yet because I feel that they are still useful, as explained in the comments (in particular to make `coq-doc.install` an interesting target). If you insist, I can do it though.

**Comment by ejgallego** [comment]
> Updated to take your comments into account. I did not comment out the install rules yet because I feel that they are still useful, as explained in the comments (in particular to make `coq-doc.install` an interesting target). If you insist, I can do it though.

That's fine to me @Zimmi48 ; just to be aware that if we try to use the opam files Dune will segfaul tho; indeed we should report upstream so at least fails with a nicer error [I think that particular case is not reported indeed]

**Comment by Zimmi48** [comment]
We should definitely report this segfault. Note that it is not clear to me that using the opam file will make Dune segfault though as the install step of opam file is not handled with `dune install` but with `opam-installer`.

**Comment by Zimmi48** [comment]
FTR `opam-installer` fails, but properly:
```
[ERROR] Could not find "/home/theo/git/coq/_build/install/default/doc/coq-doc/html/refman"
[ERROR] Could not find "/home/theo/git/coq/_build/install/default/doc/coq-doc/html/stdlib"
[ERROR] Could not find "/home/theo/git/coq/_build/install/default/doc/coq-doc/pdf/refman"
```

**Comment by Zimmi48** [comment]
Instead of generating the proper `.install`, I can override the `install` phase with a simple shell script copying the folders.

File: dev/dune (hunk 5 above)
Comment (Zimmi48) [reply dev/dune]: FTR it also happens when trying to install through opam:

```
### output ###
# File "dev/dune", line 39, characters 6-44:
# 39 |     %{lib:coq.top_printers:top_printers.cmi}
#            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: Library "coq.top_printers" not found.
# Hint: try: dune external-lib-deps --missing -p coq-doc -j 3 @install
```

**Comment by ejgallego** [comment]
> If you insist, I can do it though.

Up to you @Zimmi48 , if not further comment I will merge as-is. Whatever we do the package doesn't sadly work yet [maybe we can fix that for Dune 2.5 tho]

File: .gitlab-ci.yml (hunk 2 above)
Comment (cpitclaudel) [reply .gitlab-ci.yml]: Even when I try to build the same target twice I don't get cache reuse, because I see `[config changed ('coq_config')]`.  I wonder why that is.

Comment (ejgallego) [reply .gitlab-ci.yml]: I dunno, `coq_config` is generated by configure; but configure should not be called again in a rebuild.

If you call configure it will indeed add the build date etc... this is already scheduled to be fixed by moving to the `Build_info` lib.

Comment (cpitclaudel) [reply .gitlab-ci.yml]: Got it, I think. Let me prepare a PR

Comment (cpitclaudel) [reply .gitlab-ci.yml]: I think this PR should help: https://github.com/coq/coq/pull/11892

Comment (cpitclaudel) [reply .gitlab-ci.yml]: (as long as the two builds share the same cache, of course)

Comment (cpitclaudel) [reply .gitlab-ci.yml]: > I dunno, coq_config is generated by configure; but configure should not be called again in a rebuild.

That's not the issue: the issue is that we put the whole Python module as a cache key, and equality on that isn't stable.
